util-images/network/netperfbenchmark/pkg/worker/worker.go

```diff
@@ -0,0 +1,284 @@
+			klog.Error("Error collecting metrics:", stat)
+			reply.Error = "metrics collection failed:" + stat
+			w.createResponse(reply, &res)
+			return
+		}
+		klog.Info("Metrics collected")
+	default:
+		klog.Info("Metric collection in progress")
+		reply.Error = "metric collection in progress"
+		w.createResponse(reply, &res)
+		return
+	}
+	o, err := util.ParseResult(w.tstProtocol, w.result, w.tstDuration)
+	if err != nil {
+		reply.Error = err.Error()
+		w.createResponse(reply, &res)
+		return
+	}
+	reply.Result = o
+	//startime and other variables below can be removed in future, for debugging during initial runs
+	stStr := strconv.FormatInt(w.startedAt, 10)
+	ftStr := strconv.FormatInt(w.futureTime, 10)
+	dtStr := strconv.FormatInt(w.futureTime-w.startedAt, 10)
+	reply.WorkerStartTime = "StartedAt:" + stStr + " FutureTime:" + ftStr + " diffTime:" + dtStr
+	w.createResponse(reply, &res)
+}
+
+func (w *Worker) createResponse(resp interface{}, wr *http.ResponseWriter) {
+	klog.Info("Inside create resp")
+	(*wr).Header().Set("Content-Type", "application/json")
+	(*wr).WriteHeader(http.StatusOK)
+	klog.Info("Marshalled Resp:", resp)
+	b, err := json.Marshal(resp)
+	if err != nil {
+		klog.Info("Error marshalling to json:", err)
+	}
+	(*wr).Write(b)
+}
+
+// StartTCPClient starts iperf client for tcp measurements
+func (w *Worker) StartTCPClient(res http.ResponseWriter, req *http.Request) {
+	klog.Info("In StartTCPClient")
+	w.startWork(&res, req, util.ProtocolTCP, "iperf", tcpClientCmd)
+}
+
+// StartTCPServer starts iperf server for tcp measurements
+func (w *Worker) StartTCPServer(res http.ResponseWriter, req *http.Request) {
+	klog.Info("In StartTCPServer")
+	w.startWork(&res, req, util.ProtocolTCP, "iperf", tcpServerCmd)
+}
+
+// StartUDPServer starts iperf server for udp measurements
+func (w *Worker) StartUDPServer(res http.ResponseWriter, req *http.Request) {
+	klog.Info("In StartUDPServer")
+	w.startWork(&res, req, util.ProtocolUDP, "iperf", udpServerCmd)
+}
+
+// StartUDPClient starts iperf server for udp measurements
+func (w *Worker) StartUDPClient(res http.ResponseWriter, req *http.Request) {
+	klog.Info("In StartUDPClient")
+	w.startWork(&res, req, util.ProtocolUDP, "iperf", udpClientCmd)
+}
+
+// StartHTTPServer starts an http server for http measurements
+func (w *Worker) StartHTTPServer(res http.ResponseWriter, req *http.Request) {
+	klog.Info("In StartHTTPServer")
+	go w.startListening(httpPort, map[string]func(http.ResponseWriter, *http.Request){"/test": w.Handler})
+	w.createResponse(startWorkResponse{}, &res)
+}
+
+// StartHTTPClient starts an http client for http measurements
+func (w *Worker) StartHTTPClient(res http.ResponseWriter, req *http.Request) {
+	klog.Info("In StartHTTPClient")
+	w.startWork(&res, req, util.ProtocolHTTP, "siege", httpClientCmd)
+}
+
+// Handler handles http requests for http measurements
+func (w *Worker) Handler(res http.ResponseWriter, req *http.Request) {
+	fmt.Fprintf(res, "hi\n")
```

Comment:
Writing locally seems strange - you should write to response writer?

Reply:
Modified.